File: src/main/java/be/angelcorp/libs/celest/time/HourMinSec.java
/**
 * Copyright (C) 2009-2012 simon <simon@angelcorp.be>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *        http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.libs.celest.time;

import javax.annotation.concurrent.Immutable;

import be.angelcorp.libs.math.linear.Vector3D;

/**
 * 
 * Represents an epoch in the format of hh:mm:ss.ss, more information in {@link IHourMinSec}
 * 
 * @author Simon Billemont
 * @see IHourMinSec
 */
@Immutable
public class HourMinSec implements IHourMinSec {

	/**
	 * Integer number of hours closest to, but just before the epoch [hour]
	 */
	private final int		hour;
	/**
	 * Integer number of minutes closest to, but just before the epoch since the last complete hour
	 * [minute]
	 */
	private final int		minute;
	/**
	 * Number of seconds to the epoch from the preceding whole minute
	 */
	private final double	second;

	/**
	 * Create a HMS time based on known DMS time
	 * 
	 * @param dms
	 *            Known {@link IDegreeMinSec} time
	 */
	public HourMinSec(IDegreeMinSec dms) {
		// TODO: make it the seconds more accurate then an integer
		double rad = dms.getRadian();
		int[] arr = TimeUtils.rad_hms(rad);
		this.hour = arr[0];
		this.minute = arr[1];
		this.second = arr[2];
	}

	/**
	 * Create a hms time based on a given integer hour and minute and float second
	 * 
	 * @param hour
	 *            Hour of the epoch
	 * @param minute
	 *            Minute of the epoch
	 * @param second
	 *            Second of the epoch
	 */
	public HourMinSec(int hour, int minute, double second) {
		this.hour = hour;
		this.minute = minute;
		this.second = second;
	}

	@Override
	public double getDayFraction() {
		return (((second / 60. + minute) / 60.) + hour) / 24.;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public int getHour() {
		return hour;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public int getMinute() {
		return minute;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public double getRadian() {
		return TimeUtils.hms_rad(hour, minute, second);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public double getSecond() {
		return second;
	}

	/**
	 * Get the time as a vector
	 * 
	 * @return A vector containing <hour, min, sec>
	 */
	public Vector3D getTime() {
		return new Vector3D(hour, minute, second);
	}

	/**
	 * Represents the HMS epoch as a string in the form of 'hh:mm:ss.sss'
	 */
	@Override
	public String toString() {
		return String.format("%d:%d:%f", hour, minute, second);
	}

}

File: celest/src/main/java/be/angelcorp/celest/time/HourMinSec.scala
/**
 * Copyright (C) 2009-2012 simon <simon@angelcorp.be>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.celest.time

/**
 * Represents an epoch with a given hour, minute and seconds, from a known event, like the start of a day.
 *
 * @param hour Integer number of hours closest to, but just before the epoch [hour]
 * @param minute Integer number of minutes closest to, but just before the epoch since the last complete hour [minute]
 * @param second Number of seconds to the epoch from the preceding whole minute [sec]
 *
 * @author Simon Billemont
 */
case class HourMinSec(hour: Int, minute: Int, second: Double) {

  /**
   * Get the amount of days represented by this set of object. For example:
   * <ul>
   * <li>h=0, m=0, s=0 &rarr; 0</li>
   * <li>h=1, m=0, s=0 &rarr; 1/24</li>
   * <li>h=24, m=0, s=0 &rarr; 1</li>
   * </ul>
   *
   * @return Fraction of a Julian day represented by this object.
   */
  def dayFraction = (((second / 60.0 + minute) / 60.0) + hour) / 24.0

  /**
   * Get the equivalent amount of radians as this HMS. 24h == 360deg == 2&pi; rad
   *
   * @return Radians to the epoch [rad].
   */
  def rad = TimeUtils.hms_rad(hour, minute, second)

  /** Represents the HMS epoch as a string in the form of 'hh:mm:ss.sss' */
  override def toString = "%d:%d:%f".format(hour, minute, second)

}

object HourMinSec {

  /**
   * Create a HMS time based on known DMS time.
   *
   * @param dms Degrees, minutes and sections time
   */
  def apply(dms: DegreeMinSec) = {
    val arr = TimeUtils.rad_hms(dms.rad)
    new HourMinSec(arr._1, arr._2, arr._3)
  }

}
